lib/connection.js

```diff
@@ -13,224 +12,276 @@ const UTP_ERRORS = [
   'UTP_UNKNOWN'
 ]
 
-module.exports = Connection
-
-function Connection (utp, port, address, handle, halfOpen) {
-  stream.Duplex.call(this)
-
-  this.remoteAddress = address
-  this.remoteFamily = 'IPv4'
-  this.remotePort = port
-  this.destroyed = false
-
-  this._index = -1
-  this._utp = utp
-  this._handle = handle || Buffer.alloc(binding.sizeof_utp_napi_connection_t)
-  this._buffer = Buffer.allocUnsafe(65536 * 2)
-  this._offset = 0
-  this._view = new Uint32Array(this._handle.buffer, this._handle.byteOffset, 2)
-  this._callback = null
-  this._writing = null
-  this._error = null
-  this._connected = false
-  this._needsConnect = !handle
-  this._timeout = null
-  this._contentSize = 0
-  this._allowOpen = halfOpen ? 2 : 1
-
-  this.on('finish', this._shutdown)
-
-  binding.utp_napi_connection_init(this._handle, this, this._buffer,
-    this._onread,
-    this._ondrain,
-    this._onend,
-    this._onerror,
-    this._onclose,
-    this._onconnect,
-    this._realloc
-  )
-
-  unordered.add(utp.connections, this)
-  if (utp.maxConnections && utp.connections.length >= utp.maxConnections) {
-    utp.firewall(true)
+module.exports = class Connection extends streamx.Duplex {
+  constructor (utp, port, address, handle, halfOpen) {
+    super({
+      mapWritable: Buffer.from
```

Comment:
This copies all buffers as Buffer.from(buf) is a copy. Check if it's a string an only convert it then.

Reply:
This was a quick fix that I forgot to remove, thanks for reminding. Thinking about it though: it may be even better to add two options to UTP: `mapWritable` and `mapReadable` with defaults going to a `String → Buffer` conversion? Related: https://github.com/streamxorg/streamx/issues/47